### DISCUSSION TITLE
Is it possible to install MXNet on M1 Mac ?

### DISCUSSION THREAD

**Discussion by tetuomi** [apache/incubator-mxnet#20650] [Q&A]
I want to install mxnet on M1 Mac(OS 12.0).
if it is possible, please tell me how to install MXNet.

**Comment by algorithmy1** [comment]
following

**Comment by weekstudy** [comment]
Do you sucessed?

**Comment by LewsTherin511** [comment]
Interested in this as well.

I've been trying to install MxNet and GluonCV on a M1 Mac, and I'm using Miniconda, which supposedly supports arm-based packages installations.
I tried a couple of approaches so far, without any success

**Installation with pip**
I created the env by:

    conda create --name mc_mxnet_py39 python=3.9

and installed mxnet by:

    pip install --upgrade mxnet

I also tried specifying

    CONDA_SUBDIR=osx-arm64 pip install --upgrade mxnet

But the result is always the same, when importing mxnet I get the error:

    OSError: dlopen(/Users/cdemasi/miniconda3/envs/mc_mxnet_py39/lib/python3.9/site-packages/mxnet/libmxnet.so, 0x0006): tried: '/Users/cdemasi/miniconda3/envs/mc_mxnet_py39/lib/python3.9/site-packages/mxnet/libmxnet.so' (not a mach-o file)

During the installation, I see that it's doing:

    Collecting mxnet
    Using cached mxnet-1.6.0-py2.py3-none-any.whl (68.7 MB)

So, I assume there's an error in the mxnet-1.6.0-py2.py3-none-any.whl file, which is actually not a Mac-compatible version?


**Then, I tried building it from source**

- **_MxNet 2.0.0_**
I followed the instructions [on this page](https://mxnet.apache.org/versions/1.9.0/get_started/osx_setup), which seemed to work, I built and installed MxNet 2.0.0.
My problem now is to install GluonCV as well. I tried installing it with pip, and everything seemed to work as well. But then when , when running a simple object detection script, I got the error:

    RuntimeError: Legacy mxnet==2.0.0 detected, some modules may not work properly. mxnet>=1.4.0,<2.0.0 is required. You can use pip to upgrade mxnet `pip install -U 'mxnet>=1.4.0,<2.0.0'` or `pip install -U 'mxnet-cu100>=1.4.0,<2.0.0'`

(I also tried building Gluoncv from source as well, but I'm getting the same error, the issue seems to be the MxNet version)

-  building **_MxNet 1.7.0_** fails, when running "cmake --build ." I got this error:

    *** [3rdparty/mkldnn/src/cpu/CMakeFiles/dnnl_cpu.dir/jit_utils/jitprofiling/jitprofiling.c.o] Error 1

- I'll try building version 1.6.0, and see if it works

**Reply by TobiasRafael** [reply]
Same problem here

**Comment by harshraj22** [comment]
Following

**Comment by marcontk** [comment]
I was able to install it, previously I had already installed Xcode, macports, gfortran, cmake... these are the general steps.  

With macports: 
sudo port install ninja ccache opencv3

git clone --recursive https://github.com/apache/mxnet

cd mxnet
cp config/darwin.cmake config.cmake
Edit config.cmake and add in common libraries: set(OpenCV_DIR /opt/local/libexec/opencv3/cmake)

#---------------------------------------------
  Common libraries
#---------------------------------------------
set(USE_BLAS "apple" CACHE STRING "BLAS Vendor")

set(OpenCV_DIR /opt/local/libexec/opencv3/cmake)
set(USE_OPENCV ON CACHE BOOL "Build with OpenCV support")
set(OPENCV_ROOT "" CACHE BOOL "OpenCV install path. Supports autodetection.")

set(USE_OPENMP OFF CACHE BOOL "Build with Openmp support")
set(USE_ONEDNN ON CACHE BOOL "Build with oneDNN support")
set(USE_LAPACK ON CACHE BOOL "Build with lapack support")
set(USE_TVM_OP OFF CACHE BOOL "Enable use of TVM operator build system.")
#----------------------------------------------

Inside mxnet directory
mkdir build
cd build
cmake ..
cmake --build .
sudo make install   ( files are installed under /usr/local/lib /usr/local/include )


Obs: Using the source apache-mxnet-src-1.9.1-incubating.tar.gz I had no success.

**Comment by frank-stonybrook** [comment]
Was able to compile mxnet from scratch using the following **config.cmake.txt** file (stripping out the .txt extension when compiling) and pip install the python binding without any problems. However, when I tried to import mxnet from python (`python -c "import mxnet"`), I got following error complaining that the code was incompatible with my ARM-based M1. 

Very strange! Could someone help me? Thanks!

<img width="1343" alt="Screen Shot 2023-04-05 at 4 20 44 PM" src="https://user-images.githubusercontent.com/10855977/230202764-e2255454-7508-4a67-8464-56b84e65fbd5.png">

[config.cmake.txt](https://github.com/apache/mxnet/files/11162830/config.cmake.txt)
